packages/cloud_firestore/cloud_firestore/lib/src/utils/codec_utility.dart

```diff
@@ -0,0 +1,47 @@
+part of cloud_firestore;
+
+class _CodecUtility {
+  static Map<String, dynamic> _replaceValueWithDelegatesInMap(
+      Map<String, dynamic> data) {
+    Map<String, dynamic> output = Map.from(data);
+    output.updateAll((_, value) => _valueEncode(value));
+    return output;
+  }
+
+  static List<dynamic> _replaceValueWithDelegatesInArray(List<dynamic> data) {
+    return List.from(data).map((value) => _valueEncode(value));
+  }
+
+  static Map<String, dynamic> _replaceDelegatesWithValueInMap(
+      Map<String, dynamic> data) {
+    Map<String, dynamic> output = Map.from(data);
+    output.updateAll((_, value) => _valueDecode(value));
+    return output;
+  }
+
+  static List<dynamic> _replaceDelegatesWithValueInArray(List<dynamic> data) {
+    return List.from(data).map((value) => _valueDecode(value));
+  }
+
+  static dynamic _valueEncode(dynamic value) {
+    if (value is DocumentReference) {
+      return value._delegate;
+    } else if (value is List) {
+      _replaceValueWithDelegatesInArray(value);
```

Comment:
Since you're cloning the list in the _replaceBlah method (`List.from(data)`...) you can `return` the call to `_replaceValueWithDelegates` here. (Same with the Map)